server/linter.ts

```diff
@@ -0,0 +1,71 @@
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import {
+  IRuleResult,
+  Spectral,
+} from '@stoplight/spectral';
+import { parseYaml } from '@stoplight/spectral/dist/parsers';
+import { IParsedResult } from '@stoplight/spectral/dist/document';
+import { IRuleset } from '@stoplight/spectral/dist/types/ruleset';
+import { getLocationForJsonPath } from '@stoplight/yaml';
+import { URI } from 'vscode-uri';
+import { refResolver } from './resolver';
+import { registerFormats } from './formats';
+
+/**
+ * Wrapper for the Spectral linter that runs linting against VS Code document
+ * content in a manner similar to the Spectral CLI.
+ */
+export class Linter {
+  private spectral = new Spectral({ resolver: refResolver });
+
+  /**
+   * Initializes a new instance of the linter.
+   */
+  constructor() {
+    registerFormats(this.spectral);
+  }
+
+  /**
+   * Executes Spectral linting against a VS Code document.
+   * @param {TextDocument} document - The document to lint/validate.
+   * @param {IRuleset|undefined} ruleset - The ruleset to use during validation, if any.
+   * @return {Promise<IRuleResult[]>} The set of rule violations found. If no violations are found, this will be empty.
+   */
+  public async lint(document: TextDocument, ruleset: IRuleset | undefined): Promise<IRuleResult[]> {
+    // Unclear if we may have issues changing the ruleset on the shared Spectral
+    // instance here. If so, we may need to store a Spectral instance per
+    // document rather than using a single shared one via Linter.
+    if (ruleset) {
+      this.spectral.setRuleset(ruleset);
+    } else {
+      // No ruleset, so clear everything out.
+      this.spectral.setRuleset({
+        functions: {},
+        rules: {},
+        exceptions: {},
+      });
+    }
+
+    // It's unclear why JSON and YAML both get parsed as YAML, but that's how Spectral does it, sooooooo...
+    const text = document.getText();
+    const spec = parseYaml(text);
+
+    // There's a bug in how `json-ref-resolver` handles file:// URLs - it fails
+    // to decode any URL-encoded items because it treats a file://path/to/file
+    // and /fs/path/to/file style URL as the same thing. To bypass the issue we
+    // pre-convert the file URL into a local FS path.
+    //
+    // For in-memory documents, this URI will be `untitled:Untitled-1` or
+    // something similar. The `fsPath` will be `Untitled-1`, and that seems as
+    // reasonable as anything else for in-memory validation. This only really
+    // matters when parsing the relative links in references anyway.
+    const file = URI.parse(document.uri).fsPath;
+    const parsedResult: IParsedResult = {
```

Comment:
@tillig Could you rather build a `Document` and pass it to run?

cf. https://github.com/stoplightio/spectral/blob/develop/src/cli/services/linter/linter.ts#L81-L85

Reply:
I think Document is very new, but we could add this to the tech debt notes :)

Reply:
You're probably going to find lots of things like this. The [Javascript usage docs](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/javascript.md?srn=gh/stoplightio/spectral/docs/guides/javascript.md) only took me so far and they don't mention stuff like `httpAndFileResolver` or other little things, and the Spectral code itself has _literally no documentation_ in it at all, soooo... lots of reverse engineering why the CLI is doing what it's doing. :) You'll likely have a raft of better ways to do what I was doing, changes that will be easier for you to make post-import.